### DISCUSSION TITLE
On BMI extensions

### DISCUSSION THREAD

**Discussion by mdpiper** [csdms/bmi#107] [General]
@gregtucker, @mcflugen, and @mdpiper spent a little time discussing BMI extensions. Here are four things we think could be addressed by the wider group.

* _Definition_. What is a BMI extension?
* _Process_. How does someone propose and submit an extension for approval; how does an extension get approved and added to the BMI?
* _Location_. Where do extensions live?
* _Sustainability_. Who creates and maintains the documentation, language bindings, and examples for an extension?

Please add thoughts or things that we missed in the comments.

This discussion can eventually be converted into an issue, then addressed with a pull request into the BMI documentation.

**Comment by rmcd-mscb** [comment]
This is a perhaps a sidebar.  A part of the discussion I was finding a bit surprising yesterday was the connection of the BMI to the "framework".  I think of the BMI as it's own entity.  For example, I could have 2 FORTRAN models wrapped in a BMI and couple them together in FORTAN, with no consideration of a framework.  Maybe I simply misunderstand :).  Not intending to be flip here but I think frameworks come and go, but the models themselves live longer lives.  Should we be considering a framework as a control to what's acceptable for a BMI extension?

**Reply by gregtucker** [reply]
Good point @rmcd-mscb. Whether we name the code that calls the BMI functions of one or more components a "framework" or a "coupler" or a "driver" or just "a script" strikes me as partly just a question of semantics. Whatever we call it, in the end it is a piece of code that calls BMI functions. I suppose one might consider a "framework" to be more than just that - with pymt, for example, there are additional utility functions provided to handle things like grid remapping if needed - but in that case, I agree, a framework isn't necessary.

It can even get hierarchical: for example, a script that uses BMI calls to couple models a, b, and c could itself be considered a model, and that (coupled) model might itself then be coupled with another to form an even higher-level model.
